src/util/simplify_expr_pointer.cpp

```diff
@@ -388,41 +388,101 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
   return true;
 }
 
+static bool simple_structural_lvalue_equal(const exprt &lhs, const exprt &rhs)
+{
+  if(lhs.id() != rhs.id())
+  {
+    return false;
+  }
+  if(lhs.id() == ID_symbol)
+  {
+    return lhs.get(ID_identifier) == rhs.get(ID_identifier);
+  }
+  else if(lhs.id() == ID_member)
+  {
+    auto lhs_member_of = to_member_expr(lhs);
+    auto rhs_member_of = to_member_expr(rhs);
+    return lhs_member_of.get_component_name() ==
+             rhs_member_of.get_component_name() &&
+           simple_structural_lvalue_equal(
+             lhs_member_of.compound(), rhs_member_of.compound());
+  }
+  return false;
+}
+
+// rewrite expression &a.b to &a when appropriate
+static exprt
+collapse_union_struct_member_access(const exprt &op, const namespacet &ns)
+{
+  if(op.id() == ID_member)
+  {
+    auto member_expr = to_member_expr(op);
+    auto structlike_type = ns.follow(member_expr.compound().type());
+    if(structlike_type.id() == ID_union)
+    {
+      // pointer to union member is the same as pointer to union
+      // see n1570 6.7.2.1
+      return collapse_union_struct_member_access(member_expr.compound(), ns);
+    }
+    else if(structlike_type.id() == ID_struct)
+    {
+      auto struct_type = to_struct_type(structlike_type);
+      // there's a member access, so there should be at least one component
+      assert(!struct_type.components().empty());
+      auto first_component_name = struct_type.components().front().get_name();
+      if(member_expr.get_component_name() == first_component_name)
+      {
+        // Pointer to first member of structure is the same as pointer to structure
+        // see n1570 6.7.2.1
+        return collapse_union_struct_member_access(member_expr.compound(), ns);
+      }
+    }
+  }
+  return op;
+}
+
+static exprt
+simplify_inequality_address_of_prepare_op(const exprt &op, const namespacet &ns)
+{
+  exprt result = op;
+  if(op.id() == ID_typecast)
```

Comment:
Consider `while(result.id() == ID_typecast)`

Reply:
@thk123 I think stacked typecasts should've been flattened by this point, but I might as well.